FABulous.py

```diff
@@ -0,0 +1,744 @@
+                    # Gen switch matrix
+                    logger.info(f"Generating switch matrix for {st}")
+                    self.fabricGen.setWriterOutputFile(
+                        f"./Tile/{t}/{st}/{st}_switch_matrix.{self.extension}")
+                    self.fabricGen.genSwitchMatrix(st)
+                    logger.info(f"Generated switch matrix for {st}")
+
+                    # Gen config mem
+                    logger.info(f"Generating configMem for tile {t}")
+                    if os.path.exists(f"./Tile/{t}/{st}/{st}_ConfigMem.csv"):
+                        logger.info(
+                            f"Found bitstream mapping file {st}_ConfigMem.csv for tile {st}")
+                    else:
+                        logger.info(f"{st}_ConfigMem.csv does not exist")
+                        logger.info(
+                            f"Generating a default configMem for {st}")
+
+                    logger.info(f"Generating ConfigMem for {st}")
+                    self.fabricGen.setWriterOutputFile(
+                        f"./Tile/{t}/{st}/{st}_ConfigMem.{self.extension}")
+                    self.fabricGen.genConfigMem(
+                        st, f"./Tile/{t}/{st}/{st}_ConfigMem.csv")
+                    logger.info(f"Generated configMem for {st}")
+
+                    # Gen tile
+                    logger.info(f"Generating subtile {st}")
+                    self.fabricGen.setWriterOutputFile(
+                        f"./Tile/{t}/{st}/{st}_tile.{self.extension}")
+                    self.fabricGen.genTile(st)
+                    logger.info(f"Generated subtile {st}")
+
+                # Gen super tile
+                logger.info(f"Generating super tile {t}")
+                self.fabricGen.setWriterOutputFile(
+                    f"./Tile/{t}/{t}_tile.{self.extension}")
+                self.fabricGen.genSuperTile(t)
+                logger.info(f"Generated super tile {t}")
+                continue
+
+            # Gen switch matrix
+            self.do_gen_switch_matrix(t)
+
+            # Gen config mem
+            self.do_gen_config_mem(t)
+
+            logger.info(f"Generating tile {t}")
+            # Gen tile
+            self.fabricGen.setWriterOutputFile(
+                f"./Tile/{t}/{t}_tile.{self.extension}")
+            self.fabricGen.genTile(t)
+            logger.info(f"Generated tile {t}")
+
+        logger.info("Tile generation complete")
+
+    def complete_gen_tile(self, text: str, *ignored):
+        return self._complete_tileName(text)
+
+    def do_gen_all_tile(self, *ignored):
+        "generate all tiles"
+        logger.info("Generating all tiles")
+        self.do_gen_tile(self.allTile)
+        logger.info("Generated all tiles")
+
+    def do_gen_fabric(self, *ignored):
+        "generate the fabric base on the loaded fabric"
+        logger.info(f"Generating fabric {self.fabricGen.fabric.name}")
+        self.do_gen_all_tile()
+        self.fabricGen.genFabric()
```

Comment:
I think a `self.fabricGen.setWriterOutputFile` call is needed before `genFabric()`. same for `genTopWrapper()` further down